### ISSUE TITLE
Allowed argument for earth not being passed appropriately

### ISSUE THREAD

**Issue by blind-contours** [jeremyrcoyle/sl3#367]
I want to fit a discrete SL where I restrict the interactions to certain sets of variables. In earth, this is done through passing a function to the allowed parameter which checks to see if the interactions in predictors and parents (two variables in a basis function) meet certain criteria and returns TRUE or FALSE. I have written a function that correctly limits the interactions when I run earth out of the box - outside of SL, but when I create a discrete SL via `full_lrn_earth_2 <- Lrnr_earth$new(linpreds = FALSE, degree = 2, allowed = restrict_intxn_earth)` the SL does not restrict the interactions. It seems this parameter is not being passed to earth appropriately?

**Comment by rachaelvp** [comment]
**tl;dr**: I figured out what was going on and I submitted a PR with a fix [here](https://github.com/tlverse/sl3/pull/370).

The issue: 
See lines 96–99 of Lrnr_earth.R, highlighted [here](https://github.com/tlverse/sl3/blob/master/R/Lrnr_earth.R#L96-99).  We can see that the function `call_with_args` is used to to get earth's fit object, shown in L99, and the function `earth_fun` is passed to `call_with_args` along with the user- and sl3- specified list of arguments, `args`. Your `allowed` argument therefore exists in `args` and is passed to `call_with_args` in L99. Upon referencing [`call_with_args`](https://github.com/tlverse/sl3/blob/master/R/utils.R#L76-L117) and its default arguments, we can see that this function will only consider the arguments in `earth_fun` as valid; all other arguments provided in `args` that are not in [`names(formals(earth_fun))`](https://github.com/tlverse/sl3/blob/master/R/utils.R#L99) will be dropped and a message will be provided. The `allowed` argument is not in `names(formals(earth_fun))`, so the message should have looked like this: 
"Learner called function earth with unknown args: allowed. These will be dropped.
Check the params supported by this learner."

Did you receive this message when you tried to train `Lrnr_earth$new()` with `allowed` argument supplied?

The solution: 
As shown on page 3 of the [earth R package documentation](https://cran.r-project.org/web/packages/earth/earth.pdf), the `allowed` argument is defined in the non-default, "fit" method of the earth function. The call to `call_with_args` needs to modified so it doesn't reject the arguments that exist in this non-default function class of earth. We can specify these additional arguments by calling `call_with_args` with `other_valid` specified. I made that change in a branch I called "earth-formals", and I created a PR [here](https://github.com/tlverse/sl3/pull/370). Assuming the checks pass (usually takes about an hour), this change will be incorporated in the devel branch of sl3. In order to install the devel branch of sl3, follow these steps:
1. Clear workspace
2. Restart R
3. Run `remotes::install_github("tlverse/sl3@devel")` (you might need to install remotes R package first)
4. Restart R again.

Afterwards, when you run `library(sl3)` the devel version of sl3 will be loaded.
